### PR TITLE
Time-reverse simulation for phased-array delay design

### LINKED ISSUE #95: Add an example for defining delays for phased array element through a "time reverse" simulation
## Describe the new feature or enhancement

A notebook example that puts a point source where you want to focus your phased array and measure delays for wavefront at the position of the elements of your array

## Describe your proposed implementation

The skull adds aberrations to to the beam propagation; phased arrays have the potential to compensate for those by having different delays for each element, but estimating these delays can be challenging. One method shared by Sergio Jiménez-Gambín and Samuel Blackman is to run a "time reverse" simulation as described above to calculate the necessary delays. This would be a great example of NDK's feature flexibility.

### PR DESCRIPTION
**Pull request by charlesbmi**
#### Introduction

https://github.com/agencyenterprise/neurotechdevkit/issues/95

The skull adds aberrations to to the beam propagation; phased arrays have the potential to compensate for those by having different delays for each element, but estimating these delays can be challenging. One method shared by Sergio Jiménez-Gambín and Samuel Blackman is to run a "time reverse" simulation to calculate the necessary delays.

A "time reverse" simulation puts a point source where you want to focus your phased array and measures delays for wavefront at the position of the elements of your array.

This pull request adds an example notebook that performs the "time reverse" simulation to calculate element delays. It then validates the calculated delays by performing a standard forward-time simulation.

#### Changes
* Add a new "time reverse" simulation notebook
* Add a `PointSource` class that implements a (hypothetical) single-point ultrasound transducer

##### Not complete:
* The plot legend for `PointSource` is still a transducer image.

#### Behavior

Run the `time_reverse_simulation.ipynb` notebook with different scenarios and target/source locations. The final `matplotlib` animation should show the ultrasound pulse focusing on the true target.

Pulsed simulation:
pulsed_simulation

Steady-state simulation:
image

### CODE REVIEW

**Comment by charlesincharge** [comment]
@d-lucena  @NewtonSander should I format this notebook as a .py file and put it in https://github.com/agencyenterprise/neurotechdevkit/tree/main/docs/examples like the other examples?

**Comment by charlesincharge** [comment]
Not too urgent, but docs build is currently blocked until: https://github.com/agencyenterprise/neurotechdevkit/pull/124 merges

Error message:
```
Unexpected failing examples:
/home/circleci/project/docs/examples/plot_time_reverse.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/circleci/project/docs/examples/plot_time_reverse.py", line 62, in <module>
    true_scenario.render_layout()
  File "/home/circleci/project/src/neurotechdevkit/scenarios/_base.py", line 894, in render_layout
    field = self.get_field_data("layer").astype(int)
  File "/home/circleci/project/src/neurotechdevkit/scenarios/_base.py", line 339, in get_field_data
    return self.problem.medium.fields[field].data
  File "/home/circleci/project/src/neurotechdevkit/scenarios/_base.py", line 110, in problem
    return self._problem
AttributeError: 'Scenario2_2D' object has no attribute '_problem'
```

**Comment by NewtonSander** [comment]
> Not too urgent, but docs build is currently blocked until: #124 merges
> 
> Error message:
> 
> ```
> Unexpected failing examples:
> /home/circleci/project/docs/examples/plot_time_reverse.py failed leaving traceback:
> Traceback (most recent call last):
>   File "/home/circleci/project/docs/examples/plot_time_reverse.py", line 62, in <module>
>     true_scenario.render_layout()
>   File "/home/circleci/project/src/neurotechdevkit/scenarios/_base.py", line 894, in render_layout
>     field = self.get_field_data("layer").astype(int)
>   File "/home/circleci/project/src/neurotechdevkit/scenarios/_base.py", line 339, in get_field_data
>     return self.problem.medium.fields[field].data
>   File "/home/circleci/project/src/neurotechdevkit/scenarios/_base.py", line 110, in problem
>     return self._problem
> AttributeError: 'Scenario2_2D' object has no attribute '_problem'
> ```

The PR was merged @charlesincharge

**Comment by charlesincharge** [comment]
Thanks @NewtonSander !

**Review by d-lucena — COMMENTED** [review]
Thanks Charles! This is looking great. My comments for now are on the notebook only.

I'll continue with the actual code review later. I'm not a huge fan of how things look with the "is_point_source" implementation, and hoping there's a cleaner version, but haven't really found a good suggestion yet.

File: docs/examples/plot_time_reverse.py
Comment (charlesincharge) [reply docs/examples/plot_time_reverse.py]: This worked in Jupyter; I wasn't sure how `mkdocs` did LaTeX. Looks like it requires a little additional config:

Just now, I added the `.js` rendering file as specified by: 
https://squidfunk.github.io/mkdocs-material/setup/extensions/python-markdown-extensions/#arithmatex

Comment (charlesincharge) [reply docs/examples/plot_time_reverse.py]: The most recent commit fixes the rendering: 
image

**Comment by d-lucena** [comment]
maybe to remove the `is_point_source`, something like this in `layers.py`? What do you think? Maybe we can chat about this in our 1:1 to see what makes sense

```python
def draw_source(ax: matplotlib.axes.Axes, source: Source, drawing_params: Optional[SourceDrawingParams]) -> None:
    """Draw a layer showing the scenario sources on top of a figure.

    This layer can be added to any scenario figure in 2D.

    Args:
        ax: the axes to draw the sources on.
        source: the Source object to be drawn.
        drawing_params: Optional parameters for drawing the source. Required if the source is not a PointSource.
    """
    if isinstance(source, PointSource):
        source_artist = create_point_source_artist(source.position)
    else:
        assert drawing_params is not None
        source_artist = create_source_drawing_artist(drawing_params, transform=ax.transData)

    ax.add_artist(source_artist)
```

File: docs/examples/plot_time_reverse.py
Comment (charlesincharge) [reply docs/examples/plot_time_reverse.py:155]: good point! just added: About 7mm error in this example.

For visualization, I just masked the steady state visualization to scale it to the intensity within the brain only; it is easier to find the max pressure that way:

image

**Comment by charlesincharge** [comment]
Yeah let's chat during the 1:1! Updating the library wasn't a main focus here, so I added the `PointSource` fairly quickly at the time and opted to follow the same pattern as `source_is_flat`. It seemed like `rendering/_source.py` was meant to abstract `Source` completely away from `rendering/layers.py` (although that doesn't mean we can't break that abstraction layer). One alternative I had also considered was also combining the `source_is_flat`/`is_point_source` and any future relevant attributes as an `Enum` that defines the kind of template image to load, e.g. `focused` `linear-phased` `point` `diverging/curvilinear`

File: src/neurotechdevkit/rendering/_source.py
Comment (charlesincharge) [inline src/neurotechdevkit/rendering/_source.py:38]: @d-lucena I swapped out the `source_is_flat` and `is_point_source` flags for an `Enum` that can carry any info needed about the high-level image-rendering style.

Along with the helper classmethod `from_source`, it makes the calling code (in `scenarios/_base.py`) a bit simpler. Let me know what you think!

**Review by d-lucena — APPROVED** [review]
This looks great, Charles! Really liked your solution for the different source rendering types!
My only comment would be about calling it `SourceType` as I feel like what would come to my mind is phased array, single element, etc when we say type. Maybe `SourceRenderType`? 

I'm approving because I think it looks great as is, but if you can think of a less ambiguous name, it would be worth changing before merging.

**Comment by charlesincharge** [comment]
Thanks, updated to `SourceRenderType`!

**Pull request merged**
